src/main/kotlin/com/ajax/cryptocurrency/threads/PriceSaver.kt

```diff
@@ -1,47 +1,34 @@
 package com.ajax.cryptocurrency.threads
 
+import com.ajax.cryptocurrency.config.SingleShotBackgroundJob
 import com.ajax.cryptocurrency.model.Cryptocurrency
 import com.ajax.cryptocurrency.repository.CryptocurrencyRepository
 import org.json.JSONObject
 import org.slf4j.Logger
 import org.slf4j.LoggerFactory
-import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.beans.factory.annotation.Value
 import org.springframework.stereotype.Component
 import java.io.BufferedReader
 import java.io.InputStreamReader
-import java.lang.Thread.currentThread
 import java.lang.Thread.sleep
 import java.net.URL
 import java.time.OffsetDateTime
 import java.time.ZoneOffset
 
+@SingleShotBackgroundJob(startDelay = 1000, maxParallelThreads = 30)
 @Component
 data class PriceSaver(
-    @Autowired private val cryptocurrencyRepository: CryptocurrencyRepository,
-    @Value("\${cryptocurrency.name}") private val cryptocurrencyNames: List<String>,
-    @Value("\${cryptocurrency.thread-sleep-time}") private val threadSleepTime: Long
+    private val cryptocurrencyRepository: CryptocurrencyRepository,
+    @Value("cryptocurrencyNames") val cryptocurrencyName: String,
```

Comment:
I think we should remove `@Value("cryptocurrencyNames") val cryptocurrencyName: String,` here.

Pass `cryptocurrencyName` to `run` function. Also, we have to remove `Runnable` implementation because `Runnable#run` doesn't accepts arguments

Reply:
removed @Value("cryptocurrencyNames") val cryptocurrencyName: String from constructor